torch/lib/THCUNN/LookupTable.cu

```diff
@@ -162,36 +162,76 @@ __global__ void cunn_LookupTable_accGradParametersKernel(
   }
 }
 
-/*
- * Keep the norm of weight smaller than maxNorm
- */
 template <typename Dtype, typename Acctype>
-struct pow_v
-{
-  Acctype normType;
-  pow_v(Dtype v) : normType(ScalarConvert<Dtype, Acctype>::to(v)) {}
-  __host__ __device__
-  Acctype operator()(const Dtype& x) const {
+__device__
+Acctype fast_pow(Dtype x, Acctype normType) {
     Acctype xA = ScalarConvert<Dtype, Acctype>::to(x);
     if (normType == 1)
```

Comment:
It's likely going to do all three branches anyways with per-lane predication, unless the compiler can figure out that the parameter is uniform across the warp (which it is, because it is an argument to the kernel). The way to do this sort of thing is put this in a struct with `int Norm` as one of the template parameters, and provide different partial specialization of the struct based on `Norm` being 1, 2, or -1 (for all other powers).